### ISSUE TITLE
Cannot get brain of mobs on 1.19.2

### ISSUE THREAD

**Issue by datatags** [GamerCoder215/MobChip#8]
### Bug Type

Console Errors

### Description

I'm unable to get the brain of a sheep mob. I'm using this code:
`EntityAI ai = BukkitBrain.getBrain((Mob) entity).getGoalAI();`
Server version:
`This server is running Paper version git-Paper-142 (MC: 1.19.2) (Implementing API version 1.19.2-R0.1-SNAPSHOT) (Git: ef0e5a6)`
I'm using MobChip shaded and relocated into my plugin.

### Expected Behavior

Return a brain rather than an error message.

### Actual Behavior

I received this error message in console:
```
java.lang.NullPointerException: Cannot read the array length because "array" is null
	at java.util.Arrays.stream(Arrays.java:5428) ~[?:?]
	at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.fromNMS(ChipUtil1_19_R1.java:1744) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.fromNMS(ChipUtil1_19_R1.java:1919) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.lambda$getGoals$0(ChipUtil1_19_R1.java:136) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at java.lang.Iterable.forEach(Iterable.java:75) ~[?:?]
	at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.getGoals(ChipUtil1_19_R1.java:136) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.bukkit.BukkitAI.updateMap(BukkitAI.java:36) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.bukkit.BukkitAI.<init>(BukkitAI.java:31) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.bukkit.BukkitBrain.getGoalAI(BukkitBrain.java:57) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.cosmetics.pets.Pet.onEquip(Pet.java:88) ~[UltraCosmetics-2.9-RELEASE.jar:?]
```

I get other, different errors from getting the brains of other mobs, should I open different issues for those or post those here as well?

**Comment by GamerCoder215** [comment]
Will try to test when I get home.

Pasting the different errors here would be good, do you get no errors with different mobs, or does fetching the brain generally not work?

**Comment by datatags** [comment]
So far it hasn't worked at all for me. I got this error when I tried with a zombie, and a nearly identical one with cat:
```
[09:39:42] [Server thread/ERROR]: Cannot cast net.minecraft.world.entity.monster.EntityZombie to java.lang.Double
[09:39:42] [Server thread/ERROR]: java.base/java.lang.Class.cast(Class.java:3921)
[09:39:42] [Server thread/ERROR]: UltraCosmetics-2.9-RELEASE.jar//be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.getObject(ChipUtil1_19_R1.java:1722)
[09:39:42] [Server thread/ERROR]: UltraCosmetics-2.9-RELEASE.jar//be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.getDouble(ChipUtil1_19_R1.java:1703)
[09:39:42] [Server thread/ERROR]: UltraCosmetics-2.9-RELEASE.jar//be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.fromNMS(ChipUtil1_19_R1.java:1923)
[09:39:42] [Server thread/ERROR]: UltraCosmetics-2.9-RELEASE.jar//be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.lambda$getGoals$0(ChipUtil1_19_R1.java:136)
[09:39:42] [Server thread/ERROR]: java.base/java.lang.Iterable.forEach(Iterable.java:75)
[09:39:42] [Server thread/ERROR]: UltraCosmetics-2.9-RELEASE.jar//be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.getGoals(ChipUtil1_19_R1.java:136)
[09:39:42] [Server thread/ERROR]: UltraCosmetics-2.9-RELEASE.jar//be.isach.ultracosmetics.shaded.mobchip.bukkit.BukkitAI.updateMap(BukkitAI.java:36)
[09:39:42] [Server thread/ERROR]: UltraCosmetics-2.9-RELEASE.jar//be.isach.ultracosmetics.shaded.mobchip.bukkit.BukkitAI.<init>(BukkitAI.java:31)
[09:39:42] [Server thread/ERROR]: UltraCosmetics-2.9-RELEASE.jar//be.isach.ultracosmetics.shaded.mobchip.bukkit.BukkitBrain.getGoalAI(BukkitBrain.java:57)
[09:39:42] [Server thread/ERROR]: UltraCosmetics-2.9-RELEASE.jar//be.isach.ultracosmetics.cosmetics.pets.Pet.onEquip(Pet.java:88)
```
Interestingly this one was just a stack trace. But when it continued on to add a custom pathfinder to the brain, I got a similar (real) exception:
```
java.lang.NullPointerException: Cannot invoke "java.lang.Double.doubleValue()" because the return value of "be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.getObject(net.minecraft.world.entity.ai.goal.PathfinderGoal, String, java.lang.Class)" is null
	at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.getDouble(ChipUtil1_19_R1.java:1703) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.fromNMS(ChipUtil1_19_R1.java:1923) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.lambda$getGoals$0(ChipUtil1_19_R1.java:136) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at java.lang.Iterable.forEach(Iterable.java:75) ~[?:?]
	at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.getGoals(ChipUtil1_19_R1.java:136) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.bukkit.BukkitAI.updateMap(BukkitAI.java:36) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.bukkit.BukkitAI.<init>(BukkitAI.java:31) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.bukkit.BukkitBrain.getGoalAI(BukkitBrain.java:57) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.cosmetics.pets.Pet.onEquip(Pet.java:88) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.cosmetics.Cosmetic.equip(Cosmetic.java:84) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.cosmetics.type.CosmeticType.equip(CosmeticType.java:93) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.cosmetics.type.PetType.equip(PetType.java:90) ~[UltraCosmetics-2.9-RELEASE.jar:?]
        ...
```

I got this error when trying a bat, parrot, and several others:
```
java.lang.ExceptionInInitializerError: null
	at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1$5.<clinit>(ChipUtil1_19_R1.java:994) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.toNMS(ChipUtil1_19_R1.java:1683) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197) ~[?:?]
	at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[?:?]
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150) ~[?:?]
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596) ~[?:?]
	at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.custom(ChipUtil1_19_R1.java:1804) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.toNMS(ChipUtil1_19_R1.java:385) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil1_19_R1.addPathfinder(ChipUtil1_19_R1.java:399) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.abstraction.ChipUtil.addPathfinders(ChipUtil.java:53) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.bukkit.BukkitAI.updateAI(BukkitAI.java:43) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.bukkit.BukkitAI.add(BukkitAI.java:113) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.bukkit.BukkitAI.put(BukkitAI.java:70) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.cosmetics.pets.Pet.onEquip(Pet.java:90) ~[UltraCosmetics-2.9-RELEASE.jar:?]
        ...
Caused by: java.lang.IllegalArgumentException: Invalid key. Must be [a-z0-9/._-]: 
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:220) ~[guava-31.0.1-jre.jar:?]
	at org.bukkit.NamespacedKey.<init>(NamespacedKey.java:52) ~[paper-api-1.19.2-R0.1-SNAPSHOT.jar:?]
	at org.bukkit.NamespacedKey.minecraft(NamespacedKey.java:145) ~[paper-api-1.19.2-R0.1-SNAPSHOT.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.ai.animation.EntityAnimation.<init>(EntityAnimation.java:38) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	at be.isach.ultracosmetics.shaded.mobchip.ai.animation.EntityAnimation.<clinit>(EntityAnimation.java:16) ~[UltraCosmetics-2.9-RELEASE.jar:?]
	... 48 more
```

**Comment by GamerCoder215** [comment]
Seems like the issue is wrapping information from NMS, and achieving casting and other errors.

I'll try testing with the other jobs you tested and work with a fix for the future.

**Comment by GamerCoder215** [comment]
> Not sure if you've looked into this at all yet but I think I've managed to fix all three of these errors in my [fork](https://github.com/datatags/MobChip/commit/dcfe617d51a1c16fa662e6b180565aea5b96fb60). I've only fixed it for the 1_19_R1 module but I'm sure similar patches could be applied to the other modules.
> 
> For the `'array' is null` error, the issue was it was having was converting an `Ingredient` to normal `ItemStack`s. It looks like the problem was that the items that `Ingredient` stores are lazy-loaded, so simply changing it to use the `getItems()` method resolved the issue as far as I can tell.
> 
> For the 'cannot cast Zombie to double' error, I changed it so that the `getObject` method would continue searching if there's a ClassCastException OR a NoSuchFieldException.
> 
> For the 'invalid key' error, I just removed `EntityAnimation.EMPTY` entirely since it didn't look like it was being used anywhere.
> 
> I could open a PR for this if you like, but again it doesn't have such fixes for the other version modules (assuming they are actually required, I haven't checked.)

Thank you!
I've looked into the code, but haven't had a chance to test any improvements yet as I've been busy with other things, but I appreciate your contribution!

A PR would be nice, I'll be willing to test your changes (if you haven't yet), as well as update these changes with the other versions as well.

Thank you, again!

**Comment by GamerCoder215** [comment]
It looks like here was the issue:

- EntityAnimation's `EMPTY`'s NamespacedKey was invalid (my part)

- Casting was an issue with how Pathfinder's NMS Fields were structured (no public methods to get them, so I used reflection), the `ZombieAttack` Pathfinder that Zombies have overrode a double field in the Attack Pathfinder (its parent), and adding that `ClassCastException` catch you mentioned **would** check the superclass, so thanks for pointing that out.

- The `getItems()` method calls a `dissolve()` method that actually makes sure that the items aren't null, this method was added in 1.17 so I made some adjustments for older versions.

Thanks again for you help, closing this now.
